patroni/ha.py

```diff
@@ -937,6 +928,32 @@ def post_recover(self):
             return 'failed to start postgres'
         return None
 
+    def cancel_initialization(self):
+        logger.info('removing initialize key after failed attempt to bootstrap the cluster')
+        self.dcs.cancel_initialization()
+        self.state_handler.stop('immediate')
+        self.state_handler.move_data_directory()
+        raise PatroniException('Failed to bootstrap cluster')
+
+    def post_bootstrap(self):
+        # bootstrap has failed if postgres is not running
+        if not self.state_handler.is_running() or self._post_bootstrap_task.result is False:
+            self.cancel_initialization()
+
+        self.keepalive()
+
+        if self._post_bootstrap_task.result is None and self.state_handler.is_leader():
+            self._async_executor.schedule('post_bootstrap')
+            self._async_executor.run_async(self.state_handler.post_bootstrap,
```

Comment:
Why do we need the ping-pong between HA -> PostgreSQL -> HA, first to run bootstrap in PostgreSQL, then to check that bootstrap didn't return False in HA, just to run post_bootstrap in PostgreSQL? Couldn't we do instead  
``return do_initialize(config) and self._configure_server_parameters() and self.start() and self.bost_bootstrap()`` at the end of ha.bootstrap?